src/NuGet.Core/NuGet.Configuration/Settings/ImmutableSettings.cs

```diff
@@ -0,0 +1,62 @@
+// All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+
+namespace NuGet.Configuration
+{
+    // Represents a wrapper for an immutable settings instance.
+    // This means that any methods invoked on this instance that try to alter it will throw.
+    internal class ImmutableSettings : ISettings
+    {
+        private readonly ISettings _settings;
+
+        internal ImmutableSettings(ISettings settings)
+        {
+            _settings = settings ?? throw new ArgumentNullException(string.Format(CultureInfo.CurrentCulture, Resources.Argument_Cannot_Be_Null_Or_Empty, nameof(settings)));
+        }
+
+        public event EventHandler SettingsChanged
+        {
+            add
+            {
+                _settings.SettingsChanged += value;
+            }
+            remove
+            {
+                _settings.SettingsChanged -= value;
+            }
+        }
+        public void AddOrUpdate(string sectionName, SettingItem item)
+        {
+            throw new NotImplementedException();
```

Comment:
Throughout this class `NotSupportedException` is more appropriate.

You'd throw a `NotImplementedException` because you haven't yet got around to providing implementation.  An instance of this exception smells like a bug.

You'd throw a `NotSupportedException` because you'll never provide an implementation (e.g.:  by design).